### PR TITLE
Consolidate task that supports the `Calamari` and `Calamari.Cloud` packages

### PR DESCRIPTION
**Pull request by droyad**
This task looks at all the referenced Calamari packages and cracks open the NuPkgs and finds all the unique files within them. It then zips up those files along with an index detailing how to reconstruct the source packages.

The idea is that we can ship standalone Calamari flavours (e.g. Calamari.Terraform) for several platforms without blowing out the size of the Octopus Server install. Just Zipping them would result in 90-150mb per flavour. We plan on having 10 or so flavours. 

With this method of consolidation, the increase for each flavour is marginal, keeping the archive size to around 150-200mb. It works because 90% of each build for a platform is the same (the runtime) as well as sharing common libs across platform builds.

Sashimi formatted packages are not yet supported.

Sample [index file](https://github.com/OctopusDeploy/ConsolidateCalamariPackagesTask/files/4455969/index.txt)

image

### CODE REVIEW

File: source/ConsolidateCalamariPackagesTask/Consolidate.cs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        public (bool result, string packageFileName) Execute(string outputDirectory, IReadOnlyList<PackageReference> packageReferences)
+        {
+            if (!Directory.Exists(outputDirectory))
+            {
+                log.Error($"The output directory {outputDirectory} does not exist");
+                return (false, null);
+            }
+
+            var packagesToScan = packageReferences
+                .Where(p => /*p.Name.StartsWith("Sashimi.") ||*/ p.Name.StartsWith("Calamari"))
```
Comment (slewis74) [inline source/ConsolidateCalamariPackagesTask/Consolidate.cs:34]: Is this comment going to stay? We'd want to be scanning Sashimi packages, yeah?

**Review by slewis74 — APPROVED** [review]
Very nice! I think there may be one change required, see comment inline.

❤️ that we have a test around it too.

File: source/ConsolidateCalamariPackagesTask/Consolidate.cs (hunk 1 above)
Comment (droyad) [reply source/ConsolidateCalamariPackagesTask/Consolidate.cs:34]: It won't, I'm working on that part right now. This at least gets Consolidated package into Server without `Calamari.Terraform`.

File: source/ConsolidateCalamariPackagesTask/ConsolidateCalamariPackagesTask.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            var packageReferences = Packages
+                .Select(p => new PackageReference
+                {
+                    Name = p.GetMetadata("Name"),
+                    Version = p.GetMetadata("Version"),
+                    ResolvedPath = p.GetMetadata("ResolvedPath"),
+                })
+                .ToArray();
+            
+            var (result, packageFilename) = new Consolidate(new MsBuildTaskLog(Log)).Execute(OutputDirectory, packageReferences);
```
Comment (johnsimons) [inline source/ConsolidateCalamariPackagesTask/ConsolidateCalamariPackagesTask.cs:36]: @droyad should you be explicitly calling `Dispose` on this?

Comment (droyad) [reply source/ConsolidateCalamariPackagesTask/ConsolidateCalamariPackagesTask.cs:36]: Indeed I should be. I'll fix that up in the next PR.

**Pull request merged**